crates/core/src/client/mod.rs

```diff
@@ -133,31 +110,22 @@ impl<P: Provider + Send + Sync> KakarotEthApi<P> for KakarotClient<P> {
         let ethereum_address: Felt252Wrapper = ethereum_address.into();
         let ethereum_address = ethereum_address.into();
 
-        let mut calldata_vec = calldata.clone().into_iter().map(FieldElement::from).collect::<Vec<FieldElement>>();
+        let mut calldata = calldata.clone().into_iter().map(FieldElement::from).collect::<Vec<_>>();
 
-        let mut call_parameters =
-            vec![ethereum_address, FieldElement::MAX, FieldElement::ZERO, FieldElement::ZERO, calldata.len().into()];
-
-        call_parameters.append(&mut calldata_vec);
-
-        let request = FunctionCall {
-            contract_address: self.kakarot_address,
-            entry_point_selector: ETH_CALL,
-            calldata: call_parameters,
-        };
-
-        let call_result: Vec<FieldElement> = self.starknet_provider.call(request, starknet_block_id).await?;
+        let result = self
+            .kakarot_contract
+            .eth_call(&self.starknet_provider, &ethereum_address, &mut calldata, &starknet_block_id)
+            .await?;
 
         // Parse and decode Kakarot's call return data (temporary solution and not scalable - will
         // fail is Kakarot API changes)
         // Declare Vec of Result
         // TODO: Change to decode based on ABI or use starknet-rs future feature to decode return
         // params
-        let return_data = decode_eth_call_return(&call_result)?;
+        let return_data = decode_eth_call_return(&result)?;
```

Comment:
could we add the decoding in the `KakarotContract` itself?

Reply:
I'm wondering if it makes sense to add this decoding everywhere. Like should it be the responsibility of the `Kakarot` contract to decode from `Vec<FieldElement>` to `Bytes`?

Reply:
I think so, when you use ethers-rs for example, you expect the `contract` object to return parsed data, not the raw bytecode. In our case, I think that eth_call is supposed to return the EVM response to a eth_call, ie a `Bytes` ? (or whatever type we use in the rest of the app to manipulate such an object)